Userland/Libraries/LibWeb/DOM/Document.cpp

```diff
@@ -1560,4 +1561,52 @@ void Document::invalidate_stacking_context_tree()
         const_cast<Painting::PaintableBox*>(paint_box)->invalidate_stacking_context();
 }
 
+void Document::check_favicon_after_loading_link_resource()
+{
+    // https://html.spec.whatwg.org/multipage/links.html#rel-icon
+    // NOTE: firefox also load favicons outside the head tag, which is against spec (see table 4.6.7)
+    auto head_element = head();
+    auto favicon_link_elements = HTMLCollection::create(*head_element, [](Element const& element) {
+        if (!is<HTML::HTMLLinkElement>(element))
+            return false;
+
+        return static_cast<HTML::HTMLLinkElement const&>(element).has_loaded_icon();
+    });
+
+    if (favicon_link_elements->length() == 0) {
+        dbgln_if(SPAM_DEBUG, "No favicon found to be used");
+        return;
+    }
+
+    // 4.6.7.8 Link type "icon"
+    //
+    // If there are multiple equally appropriate icons, user agents must use the last one declared
+    // in tree order at the time that the user agent collected the list of icons.
+    auto prefered_favicon_position = favicon_link_elements->length() - 1;
+
+    // If multiple icons are provided, the user agent must select the most appropriate icon
+    // according to the type, media, and sizes attributes.
+    //
+    // FIXME: There is no selective behavior yet for favicons.
+    do {
+        auto prefered_favicon_element = favicon_link_elements->item(prefered_favicon_position);
+
+        if (prefered_favicon_element == m_active_element)
+            return;
+
+        // If the user agent tries to use an icon but that icon is determined, upon closer examination,
+        // to in fact be inappropriate (...), then the user agent must try the next-most-appropriate icon
+        // as determined by the attributes.
+        if (static_cast<HTML::HTMLLinkElement*>(prefered_favicon_element)->load_favicon_and_use_if_window_is_active()) {
+            m_active_favicon = prefered_favicon_element;
+            return;
+        }
+        if (prefered_favicon_position <= 0)
+            break;
+        prefered_favicon_position--;
+    } while (true);
```

Comment:
Why not just put `prefered_favicon_position > 0` in the while?
Also I think this is a typo prefered -> preferred

Reply:
I have to rethink this part

Indexes can't be negative here, the compiler simply doesn't like that (of course I could dodge this issue if I have used signed numbers)

I've tried in another form too but ended up with the counter starting with the list length and doing the loop conditions at the top, but it feels weird to start the counter with an out-of-bound position.

Reply:
You can probably just go through the `favicon_link_elements` in reverse order with `for (auto& favicon_element : in_reverse(favicon_link_elements)) {...}`

Reply:
Oh actually this wont work because it is not just a vector...
If the current solution works then its fine I guess

Reply:
I'll implement it that way, it seems to be a good fit here (as long as there is no messing up with order due to preferred favicons) and I don't have to maintain the counter anymore.

Reply:
I need to digg in some other options because `HTMLCollection` doesn't seem to have begin and end functions

```
In file included from ../../Userland/Libraries/LibWeb/DOM/Document.cpp:11:
../.././AK/ReverseIterator.h: In instantiation of 'auto AK::ReverseWrapper::begin(AK::ReverseWrapper::ReverseWrapper<Container>) [with Container = AK::NonnullRefPtr<Web::DOM::HTMLCollection>&]':
../../Userland/Libraries/LibWeb/DOM/Document.cpp:1591:86:   required from here
../.././AK/ReverseIterator.h:98:74: error: 'class AK::NonnullRefPtr<Web::DOM::HTMLCollection>' has no member named 'rbegin'
   98 | auto begin(ReverseWrapper<Container> wrapper) { return wrapper.container.rbegin(); }
      |                                                        ~~~~~~~~~~~~~~~~~~^~~~~~
../.././AK/ReverseIterator.h: In instantiation of 'auto AK::ReverseWrapper::end(AK::ReverseWrapper::ReverseWrapper<Container>) [with Container = AK::NonnullRefPtr<Web::DOM::HTMLCollection>&]':
../../Userland/Libraries/LibWeb/DOM/Document.cpp:1591:86:   required from here
../.././AK/ReverseIterator.h:101:72: error: 'class AK::NonnullRefPtr<Web::DOM::HTMLCollection>' has no member named 'ren
'
  101 | auto end(ReverseWrapper<Container> wrapper) { return wrapper.container.rend(); }
      |
```

Reply:
Yeah seems like its harder than I thought. Feel free to just keep your current solution it works so it's fine.

Reply:
I could simplify the loop, I'll just do that